[RyuJit/WASM] Create a WASM-targeting RyuJit

SingleAccretion opened the pull request:
Notable build system changes:
 1) Not included in clr.alljits yet. Build with clr.wasmjit.

Notable factoring choices:
 1) Introduce some new target defines for features which WASM won't need,
    like fixed size register mask support and GCInfo based on code metadata.
 2) For this initial commit, the amount of changes has been intentionally
    minimized. In the future, more things will be moved around and ifdefed out.

Contributes to https://github.com/dotnet/runtime/issues/121208.

AaronRobinsonMSFT commented on eng/Subsets.props:
```diff
@@ -219,6 +219,7 @@
     <SubsetName Include="Clr.PalTestList" OnDemand="true" Description="Generate the list of the CoreCLR PAL tests. When using the command line, use Clr.PalTests instead." />
     <SubsetName Include="Clr.Hosts" Description="The CoreCLR corerun test host." />
     <SubsetName Include="Clr.Jit" Description="The JIT for the CoreCLR .NET runtime." />
+    <SubsetName Include="Clr.WasmJit" Description="The cross-targeting WebAssembly JIT compiler for the CoreCLR .NET runtime." />
```
Why is a new subset needed? Shouldn't this just be `-a wasm -subset clr.jit`?

kg replied:
We may want to think about how to clearly message things if we ever add a client-side (HOST_WASM) wasm jit in the future, since the current naming makes it sound like that's what this is. (I think the description is perfectly clear, though.)

It'll probably be sufficient to just name it something like 'Clr.WasmClientJit' or something if it happens, but maybe it's worth explicitly naming the current one something like 'Clr.WasmCompiler' or 'Clr.WasmAot'?

SingleAccretion committed:
Create a WASM-targeting RyuJit

Notable build system changes:
 1) Not included in clr.alljits yet. Build with clr.wasmjit.

Notable factoring choices:
 1) Introduce some new target defines for features which WASM won't need,
    like fixed size register mask support and GCInfo based on code metadata.
 2) For this initial commit, the amount of changes has been intentionally
    minimized. In the future, more things will be moved around and ifdefed
    out.

SingleAccretion committed:
Add clr.wasmjit to PR CI

Alternatively, we can just go ahead and add it to alljits right now.

SingleAccretion commented on src/coreclr/jit/targetwasm.h:
```diff
@@ -0,0 +1,284 @@
+// TODO-WASM-CQ: measure if "CSE_CONSTS" is beneficial.
+#define CSE_CONSTS               1       // Enable if we want to CSE constants
+#define EMIT_TRACK_STACK_DEPTH   1       // TODO-WASM: set to 0.
+#define EMIT_GENERATE_GCINFO     0       // We don't (can't) use code metadata-based GCInfo on WASM
```
Q: when should I use `EMIT_GENERATE_GCINFO` instead of `TARGET_WASM`?
A: use it when the code you are conditioning away relates to:
1) Tracking GCness of variables and registers in codegen.
2) Tracking GCness of variables and registers in emit.
3) Emitting GCInfo via GcInfoEncoder.

A useful mental test here is the hypothetical target of "CoreCLR interpreter (code)", which would have `HAS_FIXED_REGISTER_SET` set to `0`, but `EMIT_GENERATE_GCINFO` - `1` (it would need to use variable-size bitsets, something we don't need to deal with for WASM).

SingleAccretion commented on src/coreclr/jit/CMakeLists.txt:
```diff
@@ -653,6 +688,7 @@ add_pgo(clrjit)
 
 if (CLR_CMAKE_TARGET_ARCH_ARM64 OR CLR_CMAKE_TARGET_ARCH_AMD64)
   create_standalone_jit(TARGET clrjit_universal_arm64_${ARCH_HOST_NAME} OS universal ARCH arm64)
+  create_standalone_jit(TARGET clrjit_universal_wasm_${ARCH_HOST_NAME} OS universal ARCH wasm)
```
I went back and forth a bit on whether to use the "more proper" `wasm32` name for the target architecture. In the JIT source (and I expect in the broader /runtime source as well) `TARGET_WASM` will always mean `TARGET_WASM32 || TARGET_WASM64` - a mismatch with `_wasm` being a `TARGET_WASM32`-only Jit. On the other hand, `wasm32` would be a mismatch with `Architecture.Wasm`. Matching the `Architecture` enum seems more important.

SingleAccretion commented on src/coreclr/jit/targetwasm.h:
```diff
@@ -0,0 +1,284 @@
+// TODO-WASM-CQ: measure if "CSE_CONSTS" is beneficial.
+#define CSE_CONSTS               1       // Enable if we want to CSE constants
+#define EMIT_TRACK_STACK_DEPTH   1       // TODO-WASM: set to 0.
+#define EMIT_GENERATE_GCINFO     0       // We don't (can't) use code metadata-based GCInfo on WASM
```
> There are certainly other ways to encode GCInfo, but I think we should start with something close to what we have before inventing something different.

Good point, these comments on the define are a bit misleading, since they imply a bespoke format (or no format at all). But what's actually going to be different is not (necessarily) the format, but the place where we collect and act on the liveness information (before "WASM locals allocation"). This is where I expect to emit the GCInfo as well. It may turn out to be more convenient to serialize this information (from the spilling algorithm) into local liveness, but we can adjust the defines then.

I will reword the comments.

SingleAccretion committed:
Adjust the purpose of EMIT_GENERATE_GCINFO

AndyAyersMS reviewed (commented):
This looks like a good start to me. I want to make one more pass over it...

SingleAccretion committed:
Physical -> native

AndyAyersMS committed:
Merge branch 'main' into WasmJit

AndyAyersMS reviewed (approved):
Looked again and think this is good enough for us to get started.

SingleAccretion commented on src/coreclr/jit/emit.h:
```diff
@@ -2138,7 +2156,28 @@ class emitter
+// a read,write or modify from stack location, possible def to use latency from L0 cache
+#define PERFSCORE_LATENCY_RD_STACK         PERFSCORE_LATENCY_3C
+#define PERFSCORE_LATENCY_WR_STACK         PERFSCORE_LATENCY_1C
+#define PERFSCORE_LATENCY_RD_WR_STACK      PERFSCORE_LATENCY_3C
```
No, I would expect we would only used this for instructions that store/read from the linear memory stack.

SingleAccretion commented on src/coreclr/jit/flowgraph.cpp:
```diff
@@ -1420,15 +1420,6 @@ void Compiler::fgAddSyncMethodEnterExit()
-    // By the time EH is working, we can just enable the whole thing.
-    NYI("No support for synchronized methods");
-#endif // !FEATURE_EH
-
```
> Or are we removing this because every target has EH now?

Correct.

`FEATURE_EH` (set to `0`) is dead code that was deleted a while ago, this place was just missed in the cleanup.

SingleAccretion commented on src/coreclr/jit/emit.h:
```diff
@@ -914,6 +917,8 @@ class emitter
 #elif defined(TARGET_AMD64)
 #define ID_EXTRA_BITFIELD_BITS (20)
+#elif defined(TARGET_WASM)
+#define ID_EXTRA_BITFIELD_BITS (-4)
```
All other targets need > 32 bits to represent their fields. It so happens we only need 28 (and that's before doing the TODOs). This value is `<used bits> - 32` => `28 - 32` => `-4`. You can see I had to adjust some logic below to accept this result. But it also works "as one would expect" in case we need to enable backwards navigation, which would take a few more bits.

kg commented on src/coreclr/jit/emit.h:
```diff
@@ -2138,7 +2156,28 @@ class emitter
+// a read, write or modify from constant location, possible def to use latency from L0 cache
+#define PERFSCORE_LATENCY_RD_CONST_ADDR    PERFSCORE_LATENCY_3C
+#define PERFSCORE_LATENCY_WR_CONST_ADDR    PERFSCORE_LATENCY_1C
+#define PERFSCORE_LATENCY_RD_WR_CONST_ADDR PERFSCORE_LATENCY_3C
```
OK, so we're approximating the likelihood of a cache hit/cache miss. In that case it makes sense that const addresses and stack addresses are cheaper, thank you.

kg commented on src/coreclr/jit/lower.cpp:
```diff
@@ -8195,6 +8198,10 @@ bool Lowering::TryLowerConstIntDivOrMod(GenTree* node, GenTree** nextNode)
+#elif defined(TARGET_WASM)
+        // TODO-WASM: it is not clear this whole transformation is profitable on WASM, since it increases
+        // code size while the WASM VM should be perfectly capable of expanding the DIV/MOD by itself.
+        NYI_WASM("Lowering::TryLowerConstIntDivOrMod");
```
Good call, I agree with your reasoning. We might want to just return false here, but the first person to hit the NYI can probably do that instead of us needing to do it now.

kg commented on src/coreclr/jit/targetwasm.h:
```diff
@@ -0,0 +1,284 @@
+                                         // This is also the maximum number of registers for a MultiReg node.
+
+#define NOGC_WRITE_BARRIERS      0       // No specialized WriteBarrier JIT Helpers
+#define USER_ARGS_COME_LAST      1
```
It might be beneficial for performance for them to come first, depending on how VMs allocate wasm args to registers (whether it's the first N args in the signature, or they do some sort of smarter analysis). But that is at most a WASM-TODO imo.

jkotas replied:
> don't want to burn one of them on the generic context.

Nit: I believe that this calling convention was chosen to make implementation of instantiating stubs via hand-emitted assembly easier on Windows x86. I do not think there was a strong performance reason for it.

kg commented on src/coreclr/jit/targetwasm.h:
```diff
@@ -0,0 +1,284 @@
+#define LAST_FP_ARGREG           REG_NA
+
+#define HAS_FIXED_REGISTER_SET   0       // WASM has an unlimited number of locals/registers.
+#define REGNUM_BITS              1       // number of bits in a REG_*
```
In this case it's 1 bit because we only define a SP "register" and everything else is going to be locals, right?

SingleAccretion commented on src/coreclr/jit/targetwasm.h:
```diff
@@ -0,0 +1,284 @@
+#define MAX_RET_MULTIREG_BYTES        0  // Maximum size of a struct that could be returned in more than one register (Max is an HFA or 2 doubles)
+#define MAX_ARG_REG_COUNT             1  // Maximum registers used to pass a single argument in multiple registers.
+#define MAX_RET_REG_COUNT             1  // Maximum registers used to return a value.
+#define MAX_MULTIREG_COUNT            2  // Maximum number of registers defined by a single instruction (including calls).
```
It turns out this value cannot be `< 2` because it is used to define a C array for `GenTreeLclVar` - `regNumberSmall     gtOtherReg[MAX_MULTIREG_COUNT - 1];`.

kg commented on src/coreclr/jit/targetwasm.h:
```diff
@@ -0,0 +1,284 @@
+#define REGNUM_BITS              1       // number of bits in a REG_*
+#define REGSIZE_BYTES            TARGET_POINTER_SIZE // number of bytes in one general purpose register
+#define FP_REGSIZE_BYTES         8      // number of bytes in one FP/SIMD register
+#define FPSAVE_REGSIZE_BYTES     8       // number of bytes in one FP/SIMD register that are saved/restored, for callee-saved registers
```
I'm guessing this isn't 16 because "FP/SIMD register" is referring to architectures like arm64 where there are wide dual-purpose fp/simd registers and narrower integer registers. And for us, we will treat 16-byte SIMD values/locals separately from floats? On the other hand, if our only register (see earlier) is SP, the size of that is 4 in wasm32 and 8 in wasm64, not always 8. And SP isn't a FP/SIMD register, which would mean that we then have no FP/SIMD registers at all...

SingleAccretion replied:
This value is basically only relevant to ARM64 right now, where the calling convention specifies that for a 128 bit vector register that is callee-save, the callee only needs to preserve the lower 8 bytes. So it doesn't matter what we put here.

Since our "registers" are WASM locals, and since all WASM locals are callee-save, `8` absent SIMD128 is as sensible a number as any other.

kg reviewed (approved):
Thank you for all your hard work on this!

AndyAyersMS committed:
Merge branch 'main' into WasmJit

AndyAyersMS commented:
CI is happier overall but we crossed an GUID update and the new SPMI collection is lagging. So will hold off until we can get that running again.
